dds_web/forms.py

```diff
@@ -0,0 +1,44 @@
+"""Forms used in DDS web"""
+
+# IMPORTS ################################################################################ IMPORTS #
+
+# Standard library
+
+# Installed
+import flask_wtf
+import wtforms
+
+# Own modules
+
+# FORMS #################################################################################### FORMS #
+
+
+class RegistrationForm(flask_wtf.FlaskForm):
+    """User registration form."""
+
+    first_name = wtforms.StringField("first name", validators=[wtforms.validators.InputRequired()])
+    last_name = wtforms.StringField("last name", validators=[wtforms.validators.InputRequired()])
+    facility_name = wtforms.StringField("facility name")
+    email = wtforms.StringField(
+        "email", validators=[wtforms.validators.Email()], render_kw={"readonly": True}
+    )
+    username = wtforms.StringField(
+        "username",
+        validators=[wtforms.validators.InputRequired(), wtforms.validators.Length(min=8, max=20)],
+    )
+
+    # At least: (one lower case letter)(one upper case letter)(one digit)(special character)
+    password = wtforms.PasswordField(
+        "password",
+        validators=[
+            wtforms.validators.InputRequired(),
+            wtforms.validators.Length(min=10, max=64),
+            wtforms.validators.EqualTo("confirm", message="Passwords must match!"),
+            wtforms.validators.Regexp(
+                regex="(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[^A-Za-z0-9])",
```

Comment:
```suggestion
                regex="(?=[^a-z\n]*[a-z])(?=[^A-Z\n]*[A-Z])(?=[^\d\n]*\d)^[\S]{10,64}$",
```
1.) Here the minimum length is already checked by the regex, so the length validator could be removed (unless we want to specifically return a separate error).
2.) The regex as it currently is does not capture the string, but matches only at one position. This means the regex might return several matches if the passwords starts with multiple characters of the same type, e.g. several lower-case letters. One can nicely check this by pasting it at [Regex101](https://regex101.com). I have no idea if this matters to the `Regexp()` function. 

_Additionally, one might want to consider [to just increase password length over complexity anyway.](https://xkcd.com/936/)_

Reply:
I just had a look at the `wtforms.validators.Regexp()` and you are completely right: Both regex will do!

But just to elaborate on 2. to clarify why I suggested the change, have a look how both are matched on "passWord1!":

```
import re

regex = [
    "(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[^A-Za-z0-9])",
    "(?=[^a-z\n]*[a-z])(?=[^A-Z\n]*[A-Z])(?=[^\d\n]*\d)^[\S]{10,64}$",
]

for element in regex:
    m = re.findall(re.compile(element), "passWord1!")
    if m:
        print(m)

```

This returns:
```
['', '', '', '', '']
['passWord1!']
```
Because the password starts with "pass" (several characters of the same character class e.g. lowercase letters) the regex is also matched multiple times. `len(m)` is therefore 5 for the first and 1 for the second regex. 

In R, which I know much better than Python, it often causes an error if the length of an object changes unexpectedly because this typically also requires changing the class (type in Python) dynamically. For example, a function would usually return only one numerical value that is stored in an integer variable, but under certain conditions returns several numerical values at once, so they are coerced into a list.

@inaod568